utility/src/main/java/org/oppia/util/parser/LiTagHandler.kt

```diff
@@ -1,37 +1,171 @@
-          output.setSpan(BulletSpan(), start, output.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
-        }
+    /**
+     * Called when an opening <li> tag is encountered.
+     *
+     * Inserts an invisible [Mark] span that doesn't do any styling.
+     * Instead, [closeItem] will later find the location of this span so it knows where the opening tag was.
+     */
+    fun openItem(text: Editable)
+
+    /**
+     * Called when a closing </li> tag is encountered.
+     *
+     * Pops out the invisible [Mark] span and uses it to get the opening tag location.
+     * Then, sets a [LeadingMarginSpan] from the opening tag position to closing tag position.
+     */
+    fun closeItem(text: Editable, indentation: Int)
+  }
+
+  /**
+   * Subclass of [ListTag] for unordered lists.
+   */
+  private inner class Ul : ListTag {
+
+    override fun openItem(text: Editable) {
+      appendNewLine(text)
+      start(text, BulletListItem())
+    }
+
+    override fun closeItem(text: Editable, indentation: Int) {
+      appendNewLine(text)
+
+      getLast<BulletListItem>(text)?.let { mark ->
+        setSpanFromMark(text, mark, CustomBulletSpan(context, indentation, "•"))
       }
     }
   }
+
+  /**
+   * Subclass of [ListTag] for ordered lists.
+   */
+  private inner class Ol : ListTag {
+
+    private var index = 1
+
+    override fun openItem(text: Editable) {
+      appendNewLine(text)
+      start(text, NumberListItem(index))
+      index++
+    }
+
+    override fun closeItem(text: Editable, indentation: Int) {
+      appendNewLine(text)
+      getLast<NumberListItem>(text)?.let { mark ->
+        setSpanFromMark(text, mark, CustomBulletSpan(context, indentation, "${mark.number}."))
+      }
+    }
+  }
+
+  /**
+   * These static methods are based on the Android Html class source code.
+   * https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/text/Html.java
+   */
+  companion object {
+
+    /**
+     * Appends a new line to [text] if it doesn't already end in a new line
+     */
+    private fun appendNewLine(text: Editable) {
+      if (text.isNotEmpty() && text.last() != '\n') {
+        text.append("\n")
+      }
+    }
+
+    /**
+     * Returns the most recently added span of type [T] in [text].
+     *
+     * Invisible marking spans are inserted to record the location of opening HTML tags in the text.
+     * We do this rather than using a stack in case text is inserted and the relative location shifts around.
+     *
+     * The last span corresponds to the top of the "stack".
+     */
+    private inline fun <reified T : Mark> getLast(text: Spanned) =
+      text.getSpans(0, text.length, T::class.java).lastOrNull()
+
+    /**
+     * Pops out the invisible [mark] span and uses it to get the opening tag location.
+     * Then, sets a span from the opening tag position to closing tag position.
+     */
+    private fun setSpanFromMark(text: Spannable, mark: Mark, styleSpan: Any) {
```

Comment:
Could we generify the type of styleSpan or use a common super parent? 'Any' feels a bit weird here since it's typeless, so it'd be nice to be a bit more specific if possible from a method contract perspective.

Reply:
I am passing CustomBulletSpan if it is bullet and TextLeadingMarginSpan if it is numbered list.

Reply:
Removed customBulletspan and making use of only TextLeadingMarginSpan. Hence replaced Any with TextLeadingMarginSpan.